psutil/arch/linux/users.c

```diff
@@ -5,13 +5,134 @@
+
+    dlsym_check(handle, sd_booted);
+    dlsym_check(handle, sd_get_sessions);
+    dlsym_check(handle, sd_session_get_leader);
+    dlsym_check(handle, sd_session_get_remote_host);
+    dlsym_check(handle, sd_session_get_start_time);
+    dlsym_check(handle, sd_session_get_tty);
+    dlsym_check(handle, sd_session_get_username);
+
+    return handle;
+}
+
+static PyObject *
+psutil_users_systemd(PyObject *self, PyObject *args) {
+    char **sessions_list = NULL;
+    PyObject *py_retlist = PyList_New(0);
+    PyObject *py_tuple = NULL;
+    PyObject *py_username = NULL;
+    PyObject *py_tty = NULL;
+    PyObject *py_hostname = NULL;
+    PyObject *py_user_proc = NULL;
+    double tstamp = 0.0;
+    pid_t pid = 0;
+
+    if (py_retlist == NULL)
+        return NULL;
+    int sessions = sd_get_sessions(&sessions_list);
+    for (int i = 0; i < sessions; i++) {
+        const char *session_id = sessions_list[i];
+        py_tuple = NULL;
+        py_user_proc = NULL;
+        py_user_proc = Py_True;
+
+        char *username = NULL;
+        if (sd_session_get_username(session_id, &username) < 0)
+            goto error;
+        py_username = PyUnicode_DecodeFSDefault(username);
+        free(username);
+        if (! py_username)
+            goto error;
+
+        char *tty = NULL;
+        if (sd_session_get_tty(session_id, &tty) < 0) {
+            py_tty = PyUnicode_DecodeFSDefault("n/a");
+        } else {
+            py_tty = PyUnicode_DecodeFSDefault(tty);
+            free(tty);
+        }
+        if (! py_tty)
+            goto error;
+
+        char *hostname = NULL;
+        if (sd_session_get_remote_host(session_id, &hostname) < 0)
+            goto error;
+        py_hostname = PyUnicode_DecodeFSDefault(hostname);
+        free(hostname);
+        if (! py_hostname)
+            goto error;
+
+        uint64_t usec = 0;
+        if (sd_session_get_start_time(session_id, &usec) < 0)
+           goto error;
+        tstamp = (double)usec / 1000000.0;
+
+        if (sd_session_get_leader(session_id, &pid) < 0)
+           goto error;
+
+        py_tuple = Py_BuildValue(
+            "OOOdO" _Py_PARSE_PID,
+            py_username,              // username
+            py_tty,                   // tty
+            py_hostname,              // hostname
+            tstamp,                   // tstamp
+            py_user_proc,             // (bool) user process
+            pid                       // process id
+        );
+        if (! py_tuple)
+            goto error;
+        if (PyList_Append(py_retlist, py_tuple))
+            goto error;
+        Py_CLEAR(py_username);
+        Py_CLEAR(py_tty);
+        Py_CLEAR(py_hostname);
+        Py_CLEAR(py_tuple);
+        free (sessions_list[i]);
+    }
+    free(sessions_list);
+    return py_retlist;
+
+error:
+    Py_XDECREF(py_username);
+    Py_XDECREF(py_tty);
+    Py_XDECREF(py_hostname);
+    Py_XDECREF(py_tuple);
+    Py_DECREF(py_retlist);
+    free(sessions_list);
```

Comment:
It looks like you are supposed to check whether this is NULL before free()ing it.

Reply:
Done